### ISSUE TITLE
Fan curve hysteresis/ramping

### ISSUE THREAD

**Issue by Atemu** [ilya-zlobintsev/LACT#291]
As of 0.5.0, the fan curve appears to be quite simple in that it applies whatever value corresponds to the current temperature. Temperatures can vary quite a bit moment-to-moment which results in audible jumps in fan speed. It'd be great if temperature fluctuations could be smoothed over and/or fan speeds ramped more slowly.

**Comment by ilya-zlobintsev** [comment]
Yeah, this is something that can be improved. I'm not fully sure how exactly it should work, but it shouldn't be too difficult. I'll take a look at it eventually

**Comment by Atemu** [comment]
Now that I've used it some more, one property that I think I'd like is to quickly raise fan speeds but only lower them very slowly (as in: many seconds).

When I get into a loading screen, the GPU load drops and therefore temperature. This currently causes LACT to drop fan speeds immediately which is quite audible. The game then continues after a few seconds which causes load and temperature rise which in turn results in the fan speeds being audibly adjusted again.

If it had kept a constant speed during that time, that would have created a lot less noise and resulted in lower temperatures.

This behaviour could be achieved using a (configurable?) length of time for which the temperature must be below the next stepping level for it to actually ramp down to appropriate fan speed for that level.  
Alternatively, a very long ramp could achieve this too but I'm not 100% sure that'd be the best method because ramping itself can actually cause a lot of noise too due to it hitting a bunch of different resonant frequencies during the ramp.

My current workaround is to have the same fan speed level for a wide range of temperatures (50-90°C Tj) where the low point is ~10°C above idle. During gaming, this low point is never hit even inside loading screens as the heat sink is still hot enough; it only hits this after a a minute or so of cooling down which I don't think is a bad property either.

**Comment by ilya-zlobintsev** [comment]
This has been implemented, see #292 for more info

**Comment by Atemu** [comment]
I'm currently using discrete fan speed levels; idle, operational and "OMGITSTOOHOT" with quite instant jumps between them such that whenever my GPU is >55°C, it'll use the operational fan speed up until 95°C from which point it'll ramp up to 100% because the fan curve has clearly failed at that point.

The spindown delay is working really well for me here :)

I can now set the operational speeds much higher allowing basically silent play for light games. I'll be experimenting with that a little and wait for a release :).

**Comment by Atemu** [comment]
One thing I just noticed is that it does appear to sometimes drop the fan speed for short moments despite temperature almost always being above the limit with only very short dips below the set point.

Does it check whether temperature has actually dropped after the delay?

**Comment by ilya-zlobintsev** [comment]
> Does it check whether temperature has actually dropped after the delay?

After the delay passes, it sets the PWM according to the current temperature at that point in time.
What could be happening is: the temperature first drops slightly (by a degree or two) which causes the spindown timer to start, and then after some time you get the temperature dip that lines up with the end of the timer. 

If this is indeed what is happening, you should be able to avoid it by increasing the minimum change threshold a bit so the spindown timer doesn't start in the first place.

If you're curious about what exactly is happening, you can enable `trace` level logging in the daemon, and it will print info about the fan control decisions (though it's quite spammy). And [here](https://github.com/ilya-zlobintsev/LACT/blob/master/lact-daemon/src/server/gpu_controller/mod.rs#L479) is the code that handles it.
